src/libopensc/card-masktech.c

```diff
@@ -0,0 +1,480 @@
+	SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, rv, "APDU transmit failed - verify change PIN");
+
+	/* Build a SC_PIN_CMD_CHANGE APDU */
+	memset(&change_data, 0, sizeof(change_data));
+	change_data.cmd = SC_PIN_CMD_CHANGE;
+	change_data.pin_type = 1;
+	change_data.pin_reference = data->pin_reference;
+	/* pin1 is set to null on purpose and flag set to implicit change
+	 => if there is a pinpad reader, do not ask for pin1 */
+	change_data.pin2 = data->pin2;
+	change_data.flags = data->flags | SC_PIN_CMD_IMPLICIT_CHANGE;
+	change_data.pin2.prompt = data->pin2.prompt;
+
+	/* the APDU is overwritten to avoid requesting the INITIAL pin if a PIN PAD is used */
+	masktech_build_change_pin_apdu(card, &change_apdu, &change_data, sbuf, sizeof(sbuf));
+	change_data.apdu = &change_apdu;
+	
+	rv = iso_ops->pin_cmd(card, &change_data, tries_left);
+	SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, rv, "APDU transmit failed - chnage PIN");
+
+	return 0;
+}
+
+static int masktech_pin_cmd(sc_card_t *card,
+                            struct sc_pin_cmd_data *data,
+                            int *tries_left)
+{
+	int       rv;
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	switch(data->cmd)
+	{
+	case SC_PIN_CMD_UNBLOCK:
+		rv = masktech_pin_unblock(card, data, tries_left);
+		break;
+	case SC_PIN_CMD_CHANGE:
+		rv = masktech_pin_change(card, data, tries_left);
+		break;
+	default:
+		rv = iso_ops->pin_cmd(card, data, tries_left);
+		break;
+	}
+	return rv;
+
+	
+}
+
+static int masktech_get_serialnr(sc_card_t * card, sc_serial_number_t * serial)
+{
+	struct sc_apdu apdu;
+	unsigned char apdu_resp[SC_MAX_APDU_BUFFER_SIZE-2];
+	int rv;
+	
+	if (!serial)
+		SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, SC_ERROR_INVALID_ARGUMENTS);
+
+	/* Get smart card serial number */
+	card->cla = 0x80;
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, 0x08, 0x00, 0x00);
+	apdu.resplen = sizeof(apdu_resp);
+	apdu.resp = apdu_resp;
+
+	rv = sc_transmit_apdu(card, &apdu);
+	card->cla = 0x00;
+
+	SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, rv, "APDU transmit failed");
+	
+	if (apdu.sw1 != 0x90 || apdu.sw2 != 0x00)
+		return SC_ERROR_INTERNAL;
+
+	if (SC_MAX_SERIALNR < apdu.resplen)
+	{
+		SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, SC_ERROR_INTERNAL);
+	}
+	/* cache serial number */
+	card->serialnr.len = apdu.resplen;
+	memcpy(card->serialnr.value, apdu.resp, card->serialnr.len);
+
+	/* copy and return serial number */
+	if (serial)
+		memcpy(serial, &card->serialnr, sizeof(*serial));
+
+	SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, SC_SUCCESS);
+}
+
+
+static int masktech_card_ctl(sc_card_t * card, unsigned long cmd, void *ptr)
+{
+	sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "masktech_card_ctl()\n");
+	switch (cmd) {
+		case SC_CARDCTL_GET_SERIALNR:
+			return masktech_get_serialnr(card, (sc_serial_number_t *) ptr);
+		default:
+			return SC_ERROR_NOT_SUPPORTED;
+	}
+}
+
+static struct sc_card_driver *sc_get_driver(void)
```

Comment:
The naming currently suggests that its a global call to OpenSC, no need to copy other peoples mistakes ;-) However, it's just a suggestion. You may leave it as is.